homeassistant/components/viam/__init__.py

```diff
@@ -0,0 +1,367 @@
+            raise ServiceValidationError(
+                f"A camera entity called {camera_name} was not found",
+                translation_domain=DOMAIN,
+                translation_key="camera_entity_not_found",
+                translation_placeholders={
+                    "camera_name": camera_name,
+                },
+            )
+
+        cam = self.hass.data[cam_entity.domain].get_entity(camera_name)
+
+        if cam is None:
+            raise ServiceValidationError(
+                f"A camera for the entity {camera_name} was not found",
+                translation_domain=DOMAIN,
+                translation_key="camera_not_found",
+                translation_placeholders={"camera_name": camera_name},
+            )
+
+        return await cam.async_camera_image()
+
+    async def _get_robot_client(
+        self, robot_secret: str | None, robot_address: str | None
+    ) -> RobotClient:
+        """Check initialized data to create robot client."""
+        address = self.data.get(CONF_ADDRESS)
+        payload = self.data.get(CONF_SECRET)
+        if self.data[CONF_CREDENTIAL_TYPE] == "api-key":
+            if robot_secret is None or robot_address is None:
+                raise ServiceValidationError(
+                    "The robot location and secret are required for this connection type.",
+                    translation_domain=DOMAIN,
+                    translation_key="robot_credentials_required",
+                )
+            address = robot_address
+            payload = robot_secret
+
+        if address is None or payload is None:
+            raise ServiceValidationError(
+                "The necessary credentials for the RobotClient could not be found.",
+                translation_domain=DOMAIN,
+                translation_key="robot_credentials_not_found",
+            )
+
+        credentials = Credentials(type="robot-location-secret", payload=payload)
+        robot_options = RobotClient.Options(
+            refresh_interval=0, dial_options=DialOptions(credentials=credentials)
+        )
+        return await RobotClient.at_address(address, robot_options)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up viam from a config entry."""
+    credential_type = entry.data[CONF_CREDENTIAL_TYPE]
+    payload = entry.data[CONF_SECRET]
+    auth_entity = entry.data[CONF_ADDRESS]
+    if credential_type == "api-key":
+        payload = entry.data[CONF_API_KEY]
+        auth_entity = entry.data[CONF_API_ID]
+
+    credentials = Credentials(type=credential_type, payload=payload)
+    dial_options = DialOptions(auth_entity=auth_entity, credentials=credentials)
+    viam_client = await ViamClient.create_from_dial_options(dial_options=dial_options)
+    manager = ViamManager(hass, viam_client, entry.entry_id, dict(entry.data))
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = manager
+    manager.register_services()
```

Comment:
Services should be registered in `async_setup` and not for each config entry.

Reply:
Is this rule documented somewhere? Can you help me understand why they should be registered in `async_setup`?

The reason I ask is because I see a couple of integrations that register services in the `async_setup_entry`, i.e. Abode and Adguard. The developer documentation for custom services shows example code in which services are registered for each config entry: https://developers.home-assistant.io/docs/dev_101_services#response-data

Reply:
Thanks for pointing out that the dev docs need to be updated :)